start.py

```diff
@@ -0,0 +1,121 @@
+
+
+def main(args):
+    image = '{}/{}/topology_nodebase:{}'.format(args.registry, args.namespace or DEFAULT_NAMESPACE,
+                                                args.operating_system or DEFAULT_OPERATING_SYSTEM)
+    nodes = [Node(hostname=hostname, group='nodes', image=image)
+             for hostname in args.nodes]
+
+    kerberos_volume_dir = args.kerberos_config_directory.replace('~', expanduser('~'))
+    kdc_image = '{}/{}/topology_nodebase_kerberos:{}'.format(args.registry, args.namespace or DEFAULT_NAMESPACE,
+                                                             args.operating_system or DEFAULT_OPERATING_SYSTEM)
+    kdc_hostname = args.kdc_node[0]
+    kdc_node = Node(hostname=kdc_hostname, group='kdc', image=kdc_image,
+                    volumes=[{kerberos_volume_dir: KERBEROS_VOLUME_DIR}])
+    cluster = Cluster(kdc_node, *nodes)
+    cluster.start(args.network)
+
+    logger.info('Updating KDC configurations ...')
+    realm = cluster.network.upper()
+    # Update configurations
+    krb5_conf_data = kdc_node.get_file(KDC_KRB5_CONF_FILENAME)
+    kdc_node.put_file(KDC_KRB5_CONF_FILENAME,
+                      re.sub(r'EXAMPLE.COM', realm,
+                             re.sub(r'example.com', cluster.network,
+                                    re.sub(r'kerberos.example.com', r'{}.{}'.format(kdc_hostname, cluster.network),
+                                           krb5_conf_data))))
+    kdc_conf_data = kdc_node.get_file(KDC_CONF_FILENAME)
+    kdc_node.put_file(KDC_CONF_FILENAME,
+                      re.sub(r'EXAMPLE.COM', realm,
+                             kdc_conf_data.replace(r'[kdcdefaults]',
+                                                   '[kdcdefaults]\n max_renewablelife = 7d\n max_life = 1d')))
+    acl_data = kdc_node.get_file(KDC_ACL_FILENAME)
+    kdc_node.put_file(KDC_ACL_FILENAME, re.sub(r'EXAMPLE.COM', realm, acl_data))
+
+    kdc_commands = [
+        'kdb5_util create -s -r {realm} -P kdcadmin'.format(realm=realm),
+        'kadmin.local -q "addprinc -pw {adminpw} admin/admin@{realm}"'.format(adminpw='acladmin', realm=realm)
+    ]
+
+    logger.info('Starting KDC ...')
+    # Add the following commands before starting kadmin daemon etc.
+    if args.kerberos_principals:
+        principal_list = ['{}@{}'.format(primary, realm) for primary in args.kerberos_principals.split(',')]
+        create_principals_cmds = ['kadmin.local -q "addprinc -randkey {}"'.format(principal)
+                                  for principal in principal_list]
+        kdc_commands.extend(create_principals_cmds)
+
+        create_keytab_cmd = 'kadmin.local -q "xst -norandkey -k {} {}" '.format(KDC_KEYTAB_FILENAME,
+                                                                                ' '.join(principal_list))
+        kdc_commands.append(create_keytab_cmd)
+
+    kdc_commands.extend([
+        'krb5kdc',
+        'kadmind',
+        'authconfig --enablekrb5 --update',
+        'service sshd start',
```

Comment:
How come these services don't automatically start? If the KDC image is based on the centos images, there should be an init system starting them automatically, I'd think?

Reply:
followed from topology_cdh (0.0.1) but verified that they do start. so took it out.